docs/proposals/koordlet/20231208-support-netqos.md

```diff
@@ -0,0 +1,395 @@
+	// IngressLimit describes the maximum network bandwidth can be used in the ingress direction,
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=100
+	IngressLimit *intstr.IntOrString `json:"ingressLimit,omitempty"`
+
+	// EgressRequest describes the minimum network bandwidth guaranteed in the egress direction.
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=0
+	EgressRequest *intstr.IntOrString `json:"egressRequest,omitempty"`
+	// EgressLimit describes the maximum network bandwidth can be used in the egress direction,
+	// unit: bps(bytes per second), two expressions are supported，int and string,
+	// int: percentage based on total bandwidth，valid in 0-100
+	// string: a specific network bandwidth value, eg: 50M.
+	// +kubebuilder:default=100
+	EgressLimit *intstr.IntOrString `json:"egressLimit,omitempty"`
+}
+
+type SystemStrategy struct {
+	...
+	// TotalNetworkBandwidth indicates the overall network bandwidth, cluster manager can set this field via "slo-controller-config" configmap, 
+	// and default value just taken from /sys/class/net/${NIC_NAME}/speed, unit: Mbps
+	TotalNetworkBandwidth resource.Quantity `json:"totalNetworkBandwidth,omitempty"`
+}
+```
+
+##### pod level：
+To enable fine-grained network bandwidth control for containers in a pod, `koordinator` will sync configuration file 
+content to `/var/run/koordinator/net/pods`. and then `terway-qos` or other netqos plugin(eg: tc) will do something to 
+limit the net bandwidth that container can use. content as follows:
+```yaml
+{
+  "cgroup":"/sys/fs/cgroup/xxxx",
+  "priority":0,
+  "pod": "namespacedname",
+  "podUID":"xxx",
+  # todo: network bandwidth limiting and preemption based on port/dscp
+  "qos-config": {}
+}
+```
+
+TODO: The `qos-config` is used to define the configuration for network bandwidth limitation and preemption based on port/dscp, which may look like this:
+```yaml
+{
+    "ingress": [
+        {
+            "matchs": [{
+                "type": "ip"
+            }],
+            "actions": [
+                {
+                    "action": "qos-class",
+                    "value": "l1",
+                }
+            ],
+        }
+    ],
+    "egress": [
+         {
+            "matchs": [{
+                "type": "port",
+                "expr": "=80"
+            }],
+            "actions": [
+                {
+                    "action": "qos-class",
+                    "value": "l1",
+                },
+                {
+                    "action": "dscp",
+                    "value": "",
+                },
+                {
+                    "action": "bandwidth_min",
```

Comment:
Is it possible to expose these min/max to user via pod.resources.requests and limits?

Reply:
It doesn't just describe resource.quantity, there are some fine-grained configurations such as ports/dscp, pod.anno might be more intuitive as an extension.

Reply:
We expose min/max to user, user may want its min bandwidth request to be guranteed. This may need scheduler to make it. Otherwise, sum of pod min may exceed overall machine bandwidth?

Reply:
for scheduler, we will implement a netbandwidth plugin based on pod.anno.netqosconfig/real load

Reply:
Pods may be scheduled to nodes with temporarily lower real-time bandwidth. When the load increases, it may not be guaranteed that all Pods' min requests are guaranteed.

Reply:
We also need to implements some eviction strategy on node level.

Reply:
OK, in my opinion, it is necessary to implement some scheduler plugin to directly ensure that sum of all Pods' min requests doesn't exceed total bandwidth of a node. Eviction after pod already running may be an uncessary cost.

Just sugesstion ready for evaluation.

Reply:
`request` usually less real, so eviction based on real load is necessary

Reply:
may be we need both.

Reply:
it may be not appropriate to define as extended resource for `Network QoS`, not scheduling base on capacity.
now we DO NOT want to "implement some scheduler plugin to directly ensure that sum of all Pods' min requests doesn't exceed total bandwidth of a node" yet, only prefer network idle node is enough